tests/setup.sh

```diff
@@ -0,0 +1,296 @@
+        sed -i -e 's,/usr/bin/cri-dockerd,/usr/local/bin/cri-dockerd,' /etc/systemd/system/cri-docker.service
+        systemctl daemon-reload
+        systemctl enable --now cri-docker.socket
+        systemctl is-active --quiet cri-dockerd.socket || die "failed to set up cri-dockerd systemd service" 1
+        echo "Enabled cri-dockerd.socket systemd service."
+      fi
+
+      # Install container-network-plugins.
+      echo "Installing container-network-plugins..."
+      containernetworking_tag=$(get_latest_version_tag "containernetworking" "plugins")
+      cni_plugin_version="${containernetworking_tag}"
+      cni_plugin_tar="cni-plugins-linux-amd64-${cni_plugin_version}.tgz"
+      cni_plugin_install_dir="/opt/cni/bin"
+      cd "${workspace}"
+      ${CURL} -O "https://github.com/containernetworking/plugins/releases/download/${cni_plugin_version}/${cni_plugin_tar}"
+      mkdir -p "${cni_plugin_install_dir}"
+      tar -xf "${cni_plugin_tar}" -C "${cni_plugin_install_dir}"
+      cd - >/dev/null
+      echo "Installed container-network-plugins."
+
+      echo "Installed prerequisites for $os-$arch."
+      ;;
+      # TODO: Needs implementation.
+    "arm64" | "aarch64")
+      die "the 'install_prerequisites' option is not implemented for ${os}-${arch}" 1
+      ;;
+    *)
+      die "the 'install_prerequisites' option does not support the arch ${arch} for OS ${os}" 1
+      ;;
+    esac
+    ;;
+  "Darwin")
+    case "$arch" in
+    # TODO: Needs implementation.
+    "x86_64")
+      die "the 'install_prerequisites' option is not implemented for ${os}-${arch}" 1
+      ;;
+      # TODO: Needs implementation.
+    "arm64")
+      die "the 'install_prerequisites' option is not implemented for ${os}-${arch}" 1
+      ;;
+    *)
+      die "the 'install_prerequisites' option does not support the arch ${arch} for OS ${os}" 1
+      ;;
+    esac
+    ;;
+  *)
+    die "the 'install_prerequisites' option does not support the OS ${os}" 1
+    ;;
+  esac
+}
+
+# Installing minikube and kubectl from hosted binaries because the nix unstable channel
+# versions are usually outdated.
+install_minikube_and_kubectl() {
+  local os=${1}
+  local arch=${2}
+  local workspace=${3}
+  local kubernetes_version=${4}
+
+  dl_link_os_path=""
+  dl_link_arch_path=""
+  case "$os" in
+  "GNU/Linux")
+    dl_link_os_path="linux"
+    ;;
+  "Darwin")
+    dl_link_os_path="darwin"
+    ;;
+  *)
+    die "the 'install_kubectl' option does not support the OS ${os}" 1
+    ;;
+  esac
+  case "$arch" in
+  "x86_64")
+    dl_link_arch_path="amd64"
+    ;;
+  "arm64" | "aarch64")
+    dl_link_arch_path="arm64"
+    ;;
+  *)
+    die "the 'install_kubectl' option does not support the arch ${arch}" 1
+    ;;
+  esac
+
+  # Pulling minikube and kubectl binaries.
+  cd "${workspace}"
+  ${CURL} "https://storage.googleapis.com/minikube/releases/latest/minikube-${dl_link_os_path}-${dl_link_arch_path}" -o minikube
```

Comment:
This seems to be downloading things everytime?

Reply:
Fixed it. Minikube is pulled in via nix, that makes it idempotent. Kubectl checks for existing binary with the required version, also idempotent.